### DISCUSSION TITLE
Use of Intl.Segmenter for optional better text splitting and grapheme detection in non english languages

### DISCUSSION THREAD

**Discussion by asturur** [fabricjs/fabric.js#9462] [Ideas]
The Intl.Segmenter is part of the recent internationalization api.
In particular Segmenter is available only for Chrome and recent safari so far.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/Segmenter

Segmenter can take a language as an input, or work with auto detection and split text into words or grapheme following the correct grammar rules.

Example for **words**:

```js
const str = "hello world 門ぞぽ発動け報北断ヤオメナ願難ワオ追補な視万レっぼふ石件激っさと通事ヒ逮振2多ム木58接門じぽ。修縄チタ岩完コハヱ宏4譲シスツヱ釈変ゅはレど三題へ条95議いは岩業ノステ教話レ域済ぱだ写棋けで。員内エ部干ユヘタラ掃真ねンがド各麻ナ理世ケア氾克たう品取現反作んイ寿同ケナネ去武うら択嫌る神寺ねリよ崎義エホヌス提人ぱ講担むだぴン普乏さしのふ。";
const segmenter = new Intl.Segmenter("ja", { granularity: "word" });
const segments = segmenter.segment(str);
console.log(Array.from(segments).map((x) => x.segment).join(" | "));
```

outputs

> "hello |   | world |   | 門 | ぞ | ぽ | 発動 | け | 報 | 北 | 断 | ヤオメナ | 願 | 難 | ワ | オ | 追補 | な | 視 | 万 | レ | っ | ぼ | ふ | 石 | 件 | 激 | っ | さ | と | 通 | 事 | ヒ | 逮 | 振 | 2 | 多 | ム | 木 | 58 | 接 | 門 | じ | ぽ | 。 | 修 | 縄 | チタ | 岩 | 完 | コハヱ | 宏 | 4 | 譲 | シスツヱ | 釈 | 変 | ゅ | は | レ | ど | 三 | 題 | へ | 条 | 95 | 議 | い | は | 岩 | 業 | ノ | ステ | 教 | 話 | レ | 域 | 済 | ぱ | だ | 写 | 棋 | け | で | 。 | 員 | 内 | エ | 部 | 干 | ユヘタラ | 掃 | 真 | ね | ン | が | ド | 各 | 麻 | ナ | 理世 | ケア | 氾 | 克 | た | う | 品 | 取 | 現 | 反 | 作 | ん | イ | 寿 | 同 | ケナネ | 去 | 武 | う | ら | 択 | 嫌 | る | 神 | 寺 | ね | リ | よ | 崎 | 義 | エホヌス | 提 | 人 | ぱ | 講 | 担 | むだ | ぴ | ン | 普 | 乏 | さ | し | の | ふ | 。"

While if used for **graphemes**:

const str = "hello world 門ぞぽ発動け報北断ヤオメナ願難ワオ追補な視万レっぼふ石件激っさと通事ヒ逮振2多ム木58接門じぽ。修縄チタ岩完コハヱ宏4譲シスツヱ釈変ゅはレど三題へ条95議いは岩業ノステ教話レ域済ぱだ写棋けで。員内エ部干ユヘタラ掃真ねンがド各麻ナ理世ケア氾克たう品取現反作んイ寿同ケナネ去武うら択嫌る神寺ねリよ崎義エホヌス提人ぱ講担むだぴン普乏さしのふ。";
const segmenter = new Intl.Segmenter("ja", { granularity: "grapheme" });
const segments = segmenter.segment(str);
console.log(Array.from(segments).map((x) => x.segment).join(" | "));

outputs

> "h | e | l | l | o |   | w | o | r | l | d |   | 門 | ぞ | ぽ | 発 | 動 | け | 報 | 北 | 断 | ヤ | オ | メ | ナ | 願 | 難 | ワ | オ | 追 | 補 | な | 視 | 万 | レ | っ | ぼ | ふ | 石 | 件 | 激 | っ | さ | と | 通 | 事 | ヒ | 逮 | 振 | 2 | 多 | ム | 木 | 5 | 8 | 接 | 門 | じ | ぽ | 。 | 修 | 縄 | チ | タ | 岩 | 完 | コ | ハ | ヱ | 宏 | 4 | 譲 | シ | ス | ツ | ヱ | 釈 | 変 | ゅ | は | レ | ど | 三 | 題 | へ | 条 | 9 | 5 | 議 | い | は | 岩 | 業 | ノ | ス | テ | 教 | 話 | レ | 域 | 済 | ぱ | だ | 写 | 棋 | け | で | 。 | 員 | 内 | エ | 部 | 干 | ユ | ヘ | タ | ラ | 掃 | 真 | ね | ン | が | ド | 各 | 麻 | ナ | 理 | 世 | ケ | ア | 氾 | 克 | た | う | 品 | 取 | 現 | 反 | 作 | ん | イ | 寿 | 同 | ケ | ナ | ネ | 去 | 武 | う | ら | 択 | 嫌 | る | 神 | 寺 | ね | リ | よ | 崎 | 義 | エ | ホ | ヌ | ス | 提 | 人 | ぱ | 講 | 担 | む | だ | ぴ | ン | 普 | 乏 | さ | し | の | ふ | 。"

Applying this when available would improve by a lot text wrapping for non english languages.

It also works good with latest emoji

```js
const str = "hello w😀rld 門ぞぽ発動け報北断ヤオメナ願難😶‍🌫️🧚🏿‍♂️";
const segmenter = new Intl.Segmenter("ja", { granularity: "grapheme" });
const segments = segmenter.segment(str);
console.log(Array.from(segments).map((x) => x.segment).join(" | "));
```
> "h | e | l | l | o |   | w | 😀 | r | l | d |   | 門 | ぞ | ぽ | 発 | 動 | け | 報 | 北 | 断 | ヤ | オ | メ | ナ | 願 | 難 | 😶‍🌫️ | 🧚🏿‍♂️"

For grapheme splitting is a direct replacement that we could leverage right away with a conditional ( if Intl.Segmenter ... )

While for word splitting, it requires some rethinking because today splitWords return an array without spaces and then adds back spaces during rendering and also uses spaces for justification.

When using this and assuming mixed language, we can't make assumptions that between words there are spaces, and so we need to consider spaces as words themselves that requires us to rethink how we do the justification of space.
Indeed i m not even sure if inside non splittable grapheme considered as single words we are supposed to add spaces when justifying.

I m currently experimenting with this at work and i ll report back on what i find about about usability in fabricJS

**Comment by ShaMan123** [comment]
looks good

**Reply by ShaMan123** [reply]
I think you should reference this discussion in the relevant tickets for the community to see

**Reply by asturur** [reply]
next time i have some time to comb the issues i ll do that.

**Comment by asturur** [comment]
Some additional thoughts:
We could keep consider spaces as words and use the zero-width-space as generic infix between them.
Then we could have our justify logic work on the zero-width-space with the unwanted side effects that a space will justify twice. So a justified text that used both spaces and not spaces won't be perfect, but would be ok.

This is how google doc does:
<img width="760" alt="image" src="https://github.com/fabricjs/fabric.js/assets/1194048/43e746bc-dc15-4c61-bbc9-ddf6ac22f902">

It does not seem to have inter glyph justification

**Comment by asturur** [comment]
Baseline implementation from @Kudze in #10584
